docs: manage workspaces with kubectl (#2510)
codenvy/che-docs: modules/end-user-guide/partials/proc_creating-workspaces.adoc

[id="creating-workspaces"]
= Creating workspaces

If your use case does not permit use of the {prod-short} dashboard, you can create workspaces with {orch-name} APIs by applying custom resources to the cluster.

[NOTE]
====

> Creating new workspaces through the {prod-short} dashboard provides better user experience and configuration benefits compared to using the command line:

Reviewer: ```suggestion
Creating workspaces through the {prod-short} dashboard provides better user experience and configuration benefits compared to using the command line:
```

> * As a user, you are automatically logged in to the {orch-name} cluster.

Reviewer: ```suggestion
* As a user, you are automatically logged in to the cluster.
```

* {platforms-name} clients work automatically.
* {prod-short} and its components automatically convert the target Git repository's devfile into the `DevWorkspace` and `DevWorkspaceTemplate` custom resources on the cluster.
* Access to the workspace is secured by default with the `routingClass: che` in the `DevWorkspace` of the workspace.
* Recognition of the `DevWorkspaceOperatorConfig` configuration is managed by {prod-short}.
* Recognition of configurations in `spec.devEnvironments` specified in the `CheCluster` custom resource including:

> ** Persistent storage strategy is specified with `devEnvironments.storage`
> ** Default IDE is specified with `devEnvironments.defaultEditor`
> ** Default plugins are specified with `devEnvironments.defaultPlugins`
> ** Container build configuration is specified with `devEnvironments.containerBuildConfiguration`

Reviewer: ```suggestion
** Persistent storage strategy is specified with `devEnvironments.storage`.
** Default IDE is specified with `devEnvironments.defaultEditor`.
** Default plugins are specified with `devEnvironments.defaultPlugins`.
** Container build configuration is specified with `devEnvironments.containerBuildConfiguration`.
```

====

.Prerequisites

* An active `{orch-cli}` session with permissions to create `DevWorkspace` resources in your {orch-namespace} on the cluster. See {orch-cli-link}.

* You know the relevant {prod-short} user namespace on the cluster.
+
TIP: You can visit `pass:c,a,q[{prod-url}]/api/kubernetes/namespace` to get your {prod-short} user namespace as `name`.

> * You are on the {prod-short} user namespace on the cluster.

Reviewer: ```suggestion
* You are in the {prod-short} user namespace on the cluster.
```

+
NOTE: {prod-short} administrators who intend to create workspaces for other users must create the `DevWorkspace` custom resource in a user namespace that is provisioned by {prod-short} or by the administrator. See xref:administration-guide:configuring-namespace-provisioning.adoc[].

.Procedure

. To prepare the `DevWorkspace` custom resource, copy the contents of the target Git repository's devfile.
+
.Copied devfile contents with `schemaVersion: 2.2.0`
====
[source,yaml,subs="+quotes,+attributes"]
----
components:
  - name: tooling-container
    container:
      image: quay.io/devfile/universal-developer-image:ubi8-latest
----
====
+
TIP: For more details, see the link:https://devfile.io/docs/2.2.0/what-is-a-devfile[devfile v2 documentation].

. Create a `DevWorkspace` custom resource, pasting the devfile contents from the previous step under the `spec.template` field.
+
.A `DevWorkspace` custom resource
====
[source,yaml,subs="+quotes,+attributes,+macros"]
----
kind: DevWorkspace
apiVersion: workspace.devfile.io/v1alpha2
metadata:
  name: my-devworkspace#<1>
  namespace: user1-dev#<2>
spec:

>   routingClass: che

Reviewer: I assume this is the same for both Che and Dev Spaces?

Author: Yes, it is the same for Che and Dev Spaces

  started: true#<3>
  contributions:#<4>
    - name: ide

>       uri: pass:c,a,q[{prod-url}]/plugin-registry/main/v3/plugins/che-incubator/che-code/insiders/devfile.yaml

Reviewer: @svor, I do understand that this is just an example, but to be sure:
Is the path `/plugin-registry/main/v3/plugins/che-incubator/che-code/insiders/devfile.yaml` also applicable to Dev Spaces?
(In other words, this path is not different for Dev Spaces due to https://github.com/redhat-developer/devspaces-images/tree/devspaces-3-rhel-8/devspaces-code, correct?)

Author: @max-cx yes, the `/plugin-registry/v3/plugins/che-incubator/che-code/insiders/devfile.yaml` is also applicable to Dev Spaces, and will refer to the Dev Spaces version of Che Code.

ex: https://devspaces.apps.sandbox.x8i5.p1.openshiftapps.com/plugin-registry/v3/plugins/che-incubator/che-code/insiders/devfile.yaml

Reviewer 2: @max-cx  yes, the path is applicable to DS. It's identical in Che and DS

Reviewer: @svor, @dkwon17, and @nickboldt, thank you for confirming.

Re https://redhat-developer.github.io/devspaces/che-plugin-registry/3.5/x86_64/v3/plugins/ , as provided by Nick, which contains `"/v3/plugins/che-incubator/che-code/insiders/devfile.yaml"`.

  template:
    projects:#<5>
      - name: my-project-name
        git:
          remotes:
            origin: https://github.com/eclipse-che/che-docs
    components:#<6>
      - name: tooling-container
        container:
          image: quay.io/devfile/universal-developer-image:ubi8-latest
----
<1> Name of the `DevWorkspace` custom resource. This will be the name of the new workspace.
<2> User namespace, which is the target {orch-namespace} for the new workspace.
<3> Determines whether the workspace must be started when the `DevWorkspace` custom resource is created.
<4> URL reference to the link:https://github.com/microsoft/vscode[Microsoft Visual Studio Code - Open Source] IDE devfile from the plugin registry.
<5> Details about the Git repository to clone into the workspace when it starts.
<6> List of components such as workspace containers and volume components.
====

. Apply the `DevWorkspace` custom resource to the cluster.

> 

Reviewer: ```suggestion

.Verification
```

. Verify that the workspace is starting by checking the *PHASE* status of the `DevWorkspace`.
+
[subs="+quotes,attributes"]
----

> $ {orch-cli} get devworkspaces -n __<user_{orch-namespace}>__  --watch

Reviewer: ```suggestion
$ {orch-cli} get devworkspaces -n __<user_{orch-namespace}>__ --watch
```

----
+
.Output
====
[subs="+quotes,attributes"]
----
NAMESPACE        NAME                  DEVWORKSPACE ID             PHASE      INFO
user1-dev        my-devworkspace       workspacedf64e4a492cd4701   Starting   Waiting for workspace deployment
----
====

. When the workspace has successfully started, its *PHASE* status changes to *Running* in the output of the `{orch-cli} get devworkspaces` command.
+
.Output
====
[subs="+quotes,attributes"]
----
NAMESPACE            NAME                  DEVWORKSPACE ID             PHASE      INFO
user1-dev            my-devworkspace       workspacedf64e4a492cd4701   Running    https://url-to-workspace.com
----
====
+
You can then open the workspace by using one of these options:
+

> ** Visit the URL provided in the `INFO` section of the output of the `{orch-cli} get devworkspaces` command.

Reviewer: ```suggestion
** Visit the URL provided in the *INFO* section of the output of the `{orch-cli} get devworkspaces` command.
```

* Access to the workspace is secured by default with the `routingClass: che` in the `DevWorkspace` of the workspace.
* Recognition of the `DevWorkspaceOperatorConfig` configuration is managed by {prod-short}.
* Recognition of configurations in `spec.devEnvironments` specified in the `CheCluster` custom resource including:
** Persistent storage strategy is specified with `devEnvironments.storage`.
** Default IDE is specified with `devEnvironments.defaultEditor`.
** Default plugins are specified with `devEnvironments.defaultPlugins`.
** Container build configuration is specified with `devEnvironments.containerBuildConfiguration`.

====

.Prerequisites

* An active `{orch-cli}` session with permissions to create `DevWorkspace` resources in your {orch-namespace} on the cluster. See {orch-cli-link}.

* You know the relevant {prod-short} user namespace on the cluster.
+
TIP: You can visit `pass:c,a,q[{prod-url}]/api/kubernetes/namespace` to get your {prod-short} user namespace as `name`.

* You are in the {prod-short} user namespace on the cluster.

> +

Reviewer: ```suggestion
+
[TIP]
====
On OpenShift, you can use the command-line `oc` tool to link:https://docs.openshift.com/container-platform/latest/cli_reference/openshift_cli/developer-cli-commands.html#oc-project[display your current namespace or switch to a namespace].
====
+
```

[id="creating-workspaces"]
= Creating workspaces

If your use case does not permit use of the {prod-short} dashboard, you can create workspaces with {orch-name} APIs by applying custom resources to the cluster.

[NOTE]
====

> Using the {prod-short} dashboard is the default way to create workspaces because {prod-short} and its components automatically convert the target Git repository's devfile into the `DevWorkspace` and `DevWorkspaceTemplate` custom resources on the cluster with the following configuration benefits:

Reviewer: ```suggestion
Creating new workspaces through the {prod-short} dashboard provides better user experience and configuration benefits compared to using the command line:
```

Reviewer 3: I was expecting the "the developer is automatically logged in the Kubernetes cluster" expressed as one more bullet point in list of benefits of using the dashboard. The the bullet points are clear and effective, I love that. I would prefer that rather than a longer introduction but this is a minor thing so it's also good enough like that.

Author: In my most recent commits, I've added the bullet point:

https://github.com/eclipse-che/che-docs/blob/9175f3a1d074fe6948632cb152ac7db098373883/modules/end-user-guide/partials/proc_creating-workspaces.adoc#L12-L19

    - name: checode
      volume:
        ephemeral: true
  events:
    preStart:
      - init-container-command
----
<1> Name of the `DevWorkspaceTemplate` custom resource. This name will be used to reference the `DevWorkspaceTemplate` custom resource within the `DevWorkspace` custom resource.
<2> User namespace, which is the target {orch-namespace} for the new workspace.
====

. To prepare the `DevWorkspace` custom resource, copy the contents of the target Git repository's devfile.
+
.Copied devfile contents with `schemaVersion: 2.2.0`
====
[source,yaml,subs="+quotes,+attributes"]
----
components:
  - name: tooling-container
    container:
      image: quay.io/devfile/universal-developer-image:ubi8-latest
----
====
+
TIP: For more details, see the link:https://devfile.io/docs/2.1.0/what-is-a-devfile[devfile v2 documentation].

. Create a `DevWorkspace` custom resource, pasting the devfile contents from the previous step under the `spec.template` field.
+
.A `DevWorkspace` custom resource
====
[source,yaml,subs="+quotes,+attributes"]
----
kind: DevWorkspace
apiVersion: workspace.devfile.io/v1alpha2
metadata:
  name: my-devworkspace#<1>
  namespace: user1-dev#<2>

> spec:

Reviewer: ```suggestion
spec:
  routingClass: che
```

Author: It is working, thank you

[id="creating-workspaces"]
= Creating workspaces

If your use case does not permit use of the {prod-short} dashboard, you can create workspaces with {orch-name} APIs by applying custom resources to the cluster.

[NOTE]
====

Using the {prod-short} dashboard is the default way to create workspaces because {prod-short} and its components automatically convert the target Git repository's devfile into the `DevWorkspace` and `DevWorkspaceTemplate` custom resources on the cluster with the following configuration benefits:

> * The developer is automatically logged into the {orch-name} cluster

Reviewer: ```suggestion
* As a user, you are automatically logged in to the {orch-name} cluster.
* {platforms-name} clients work automatically.
* {prod-short} and its components automatically convert the target Git repository's devfile into the `DevWorkspace` and `DevWorkspaceTemplate` custom resources on the cluster.
```

[id="creating-workspaces"]
= Creating workspaces

If your use case does not permit use of the {prod-short} dashboard, you can create workspaces with {orch-name} APIs by applying custom resources to the cluster.

[NOTE]
====

Creating new workspaces through the {prod-short} dashboard provides better user experience and configuration benefits compared to using the command line:

* As a user, you are automatically logged in to the {orch-name} cluster.
* {platforms-name} clients work automatically.
* {prod-short} and its components automatically convert the target Git repository's devfile into the `DevWorkspace` and `DevWorkspaceTemplate` custom resources on the cluster.

> * Secures workspace access by default, using the `routingClass: che` in the workspace's `DevWorkspace`

Reviewer: ```suggestion
* Access to the workspace is secured by default with the `routingClass: che` in the `DevWorkspace` of the workspace.
```

              discoverable: false
              urlRewriteSupported: true
            exposure: public
            name: code-redirect-2
            protocol: http
            targetPort: 13132
          - attributes:
              discoverable: false
              urlRewriteSupported: true
            exposure: public
            name: code-redirect-3
            protocol: http
            targetPort: 13133
      name: vscode-runtime-description
    - name: checode
      volume:
        ephemeral: true
  events:
    preStart:
      - init-container-command
----
<1> Name of the `DevWorkspaceTemplate` custom resource. This name will be used to reference the `DevWorkspaceTemplate` custom resource within the `DevWorkspace` custom resource.
<2> User namespace, which is the target {orch-namespace} for the new workspace.
====

. To prepare the `DevWorkspace` custom resource, copy the contents of the target Git repository's devfile.
+
.Copied devfile contents with `schemaVersion: 2.2.0`
====
[source,yaml,subs="+quotes,+attributes"]
----
components:
  - name: tooling-container
    container:
      image: quay.io/devfile/universal-developer-image:ubi8-latest
----
====
+

> TIP: For more details, see the link:https://devfile.io/docs/2.1.0/what-is-a-devfile[devfile v2 documentation].

Reviewer 4: ```suggestion
TIP: For more details, see the link:https://devfile.io/docs/2.2.0/what-is-a-devfile[devfile v2 documentation].
```

. Create a `DevWorkspace` custom resource, pasting the devfile contents from the previous step under the `spec.template` field.
+
.A `DevWorkspace` custom resource
====
[source,yaml,subs="+quotes,+attributes"]
----
kind: DevWorkspace
apiVersion: workspace.devfile.io/v1alpha2
metadata:
  name: my-devworkspace#<1>
  namespace: user1-dev#<2>
spec:
  routingClass: che
  started: true#<3>
  contributions:#<4>
    - name: ide
      kubernetes:

>         name: che-code#<5>

Reviewer 4: Does it make sense to also explain that you can skip the "creating a DWT" part if your resource is available online? Contributions can also be specified by URI, e.g. 
```yaml
contributions:
  - name: ide
    uri: <URL for devfile/devworkspace content for editor>
```

Reviewer 3: This is a good idea. And also I was thinking that it will be hard to maintain the DWT sample as part of the documentation. It would be better if we could just reference the DWT of `che-incubator/che-code/insiders` from the plugin registry. But I am afraid that we do not publish it (we only publish the `devfile.yaml`). @svor is that correct?

Reviewer 2: Currently plugin reg publishes all editors as devfile.yaml files: https://eclipse-che.github.io/che-plugin-registry/main/v3/plugins/che-incubator/che-code/insiders/devfile.yaml

Reviewer 4: The URI field can be either a DevWorkspaceTemplate or Devfile, DWO handles both. This is used in the built-in DWO samples, for example: https://github.com/devfile/devworkspace-operator/blob/main/samples/theia-latest-contributions.yaml#L22

Author: Any thoughts on removing the DWT sample in the docs and replacing it with:
```
export DWT_SPEC=$(curl https://<che_fqdn>/plugin-registry/main/v3/plugins/che-incubator/che-code/insiders/devfile.yaml | yq 'del(.schemaVersion, .metadata)')

cat <<EOF | kubectl apply -f -
{
  "apiVersion":"workspace.devfile.io/v1alpha2",
  "kind":"DevWorkspaceTemplate",
  "metadata":{
    "name":"che-code"
  },
  "spec":$DWT_SPEC
}
EOF

```

so that we don't need to maintain it directly in the docs? The above is working for upstream and downstream

Reviewer 3: Can't you reference the che-code `devfile.yaml` in the devworksapce`contribution.uri` as mentioned by @amisevsk?

Reviewer 4: The only downside of this approach is that DWO will download the devfile.yaml on every reconcile (though this hasn't been a huge issue in the past). 

The benefit of referencing the devfile by URI is that it'll automatically get updates when Che is updated. With a manually-created DWT, it'll use a static UDI image unless you manually update it.

Author: @l0rd That works too, but IMHO I'm not sure if that method can replace the DWT sample in the docs, since DWT are used when creating workspaces from the dashboard and is an important workspace-related CR to know about

My new commit adds this tip, but still documents the DWT sample too:
![image](https://user-images.githubusercontent.com/83611742/214635936-b2d1b0ee-f816-447f-ae65-7700faa9a2d2.png)

Reviewer 3: In this article we are introducing the DevWorkspace CR and that's already enough from my point of view. The DWT CR and how to define an editor is another concept and can be overwhelming. It would probably be worth a separate article so I would be happy if we could hide this complexity using the URI. 

And I like your approach that uses the link to the che-code definition of the particular Che plugin registry instance. I think that's better than using the link to the upstream online plugin registry che-code definition.

Author: Sounds good, thank you for the feedback, I've updated it to use the URI method cc @max-cx

      - name: my-project-name
        git:
          remotes:
            origin: https://github.com/eclipse-che/che-docs
    components:#<7>
      - name: tooling-container
        container:
          image: quay.io/devfile/universal-developer-image:ubi8-latest
----
<1> Name of the `DevWorkspace` custom resource. This is will be the name of the new workspace.
<2> User namespace, which is the target {orch-namespace} for the new workspace.
<3> Determines whether the workspace must be started when the `DevWorkspace` custom resource is created.
<4> Reference to the `DevWorkspaceTemplate` custom resource of the selected IDE.
<5> Name of the `DevWorkspaceTemplate` custom resource from the previous step.
<6> Details about the Git repository to clone into the workspace when it starts.
<7> List of components such as workspace containers and volume components.
====

. Apply the `DevWorkspace` custom resource to the cluster.

. Verify that the workspace is starting by checking the *PHASE* status of the `DevWorkspace`.
+
[subs="+quotes,attributes"]
----
$ {orch-cli} get devworkspaces -n __<user_{orch-namespace}>__  --watch
----
+
.Output
====
[subs="+quotes,attributes"]
----

> NAMESPACE            NAME                  DEVWORKSPACE ID             PHASE      INFO
> user1-dev	     my-devworkspace       workspacedf64e4a492cd4701   Starting   Waiting for workspace deployment

Reviewer 4: It looks like you changed the namespace from something longer, breaking alignment on the table:

```suggestion
NAMESPACE        NAME                  DEVWORKSPACE ID             PHASE      INFO
user1-dev	     my-devworkspace       workspacedf64e4a492cd4701   Starting   Waiting for workspace deployment
```

Reviewer 4: Ah actually it looks like it's a mix of tabs and spaces here (the original code looks different in the view vs in the PR comment :upside_down_face:)
```suggestion
NAMESPACE        NAME                  DEVWORKSPACE ID             PHASE      INFO
user1-dev        my-devworkspace       workspacedf64e4a492cd4701   Starting   Waiting for workspace deployment
```

Reviewer: @amisevsk, :smile: that was me, sorry. Thanks for pointing this out.

----
====

. When the workspace has successfully started, its *PHASE* status changes to *Running* in the output of the `{orch-cli} get devworkspaces` command.
+
.Output
====
[subs="+quotes,attributes"]
----
NAMESPACE            NAME                  DEVWORKSPACE ID             PHASE      INFO

> user1-dev	     my-devworkspace       workspacedf64e4a492cd4701   Running    https://url-to-workspace.com

Reviewer: ```suggestion
user1-dev            my-devworkspace       workspacedf64e4a492cd4701   Running    https://url-to-workspace.com
```

[id="creating-workspaces"]
= Creating workspaces

If your use case does not permit use of the {prod-short} dashboard, you can create workspaces with {orch-name} APIs by applying custom resources to the cluster.

[NOTE]
====

Using the {prod-short} dashboard is the default way to create workspaces because {prod-short} and its components automatically convert the target Git repository's devfile into the `DevWorkspace` and `DevWorkspaceTemplate` custom resources on the cluster with the following configuration benefits:

> 

Reviewer 3: The main benefit from an devex point of view,  is that, when using the dashboard, the developer is automatically logged into the cluster. When using the command line that's not true and Kube/OpenShift clients won't work automatically.

Reviewer: @l0rd, we can incorporate what you've just described into the text. The question to you is whether we should still keep the content that is currently on line `10` ^. WDYT? (We can edit it too.)

Reviewer 3: Sure, let's keep the content that is currently on line `10`.

* Recognition of configurations in `spec.devEnvironments` specified in the `CheCluster` custom resource including:
** Persistent storage strategy specified with `devEnvironments.storage`
** Default IDE specified with `devEnvironments.defaultEditor`
** Default plugins specified with `devEnvironments.defaultPlugins`
** Container build configuration specified with `devEnvironments.containerBuildConfiguration`
* Recognition of the `DevWorkspaceOperatorConfig` configuration managed by {prod-short}

====

.Prerequisites

* An active `{orch-cli}` session with permissions to create `DevWorkspace` and `DevWorkspaceTemplate` resources in your {orch-namespace} on the cluster. See {orch-cli-link}.

* You know the relevant {prod-short} user namespace on the cluster.
+
TIP: You can visit `pass:c,a,q[{prod-url}]/api/kubernetes/namespace` to get your {prod-short} user namespace as `name`.

* You are on the {prod-short} user namespace on the cluster.
+

> [TIP]

Reviewer 3: This tip is not useful in this context (i.e. in all the commands below the namespace is explicitly specified). And, if the user wants to change the current OpenShift project, he should be able to learn it from the OpenShift documentation, I would not expect Eclipse Che documentation to explain that. @max-cx can we remove that?

Reviewer: @l0rd, sure. I've had a look at it and I'll remove it from this procedure.

> [TIP]
> ====
> On OpenShift:
> 
> * You can use the `oc` command-line tool to check your current namespace:
> +
> `$ oc project`
> 
> * You can switch to your {prod-short} user namespace on a command line if needed:
> +
> `$ oc project __<user_namespace>__`
> ====
> +

Reviewer: ```suggestion
```

          - attributes:
              discoverable: false
              urlRewriteSupported: true
            exposure: public
            name: code-redirect-1
            protocol: http
            targetPort: 13131
          - attributes:
              discoverable: false
              urlRewriteSupported: true
            exposure: public
            name: code-redirect-2
            protocol: http
            targetPort: 13132
          - attributes:
              discoverable: false
              urlRewriteSupported: true
            exposure: public
            name: code-redirect-3
            protocol: http
            targetPort: 13133
      name: vscode-runtime-description
    - name: checode
      volume:
        ephemeral: true
  events:
    preStart:
      - init-container-command
----
<1> Name of the `DevWorkspaceTemplate` custom resource. This name will be used to reference the `DevWorkspaceTemplate` custom resource within the `DevWorkspace` custom resource.
<2> User namespace, which is the target {orch-namespace} for the new workspace.
====

. To prepare the `DevWorkspace` custom resource, copy the contents of the target Git repository's devfile.
+

> .Copied devfile contents with `schemaVersion: 2.1.0`

Reviewer 3: ```suggestion
.Copied devfile contents with `schemaVersion: 2.2.0`
```

====
[source,yaml,subs="+quotes,+attributes"]
----
components:
  - name: tooling-container
    container:
      image: quay.io/devfile/universal-developer-image:ubi8-latest
----
====
+
TIP: For more details, see the link:https://devfile.io/docs/2.1.0/what-is-a-devfile[devfile v2 documentation].

> . Create a `DevWorkspace` custom resource definition, pasting the devfile contents from the previous step under the `spec.template` field.

Reviewer 3: ```suggestion
. Create a `DevWorkspace` custom resource, pasting the devfile contents from the previous step under the `spec.template` field.
```

+

> .A `DevWorkspace` custom resource definition

Reviewer 3: ```suggestion
.A `DevWorkspace` custom resource
```

Reviewer: @l0rd, I'm a bit confused. In your previous suggestion above, you suggest removing `definition` and in this suggestion you suggest adding `definition`. That's intended by you, correct? (Just double-checking.)

Reviewer 3: I am removing the word `definition` in both comments.

Reviewer: Apologies, seems I got dyslexic while looking at the GItHub diffs.

[id="creating-workspaces"]
= Creating workspaces

If your use case does not permit use of the {prod-short} dashboard, you can create workspaces with {orch-name} APIs by applying custom resources to the cluster.

> [NOTE]

Reviewer: @l0rd, I've moved all remaining points from the intro paragraph into the bullet list, WDYT?
![image](https://user-images.githubusercontent.com/78564304/214516261-6db6c03b-7861-4575-b0e7-e0f085d5020e.png)

Reviewer 3: 👍

+
.Copied devfile contents with `schemaVersion: 2.2.0`
====
[source,yaml,subs="+quotes,+attributes"]
----
components:
  - name: tooling-container
    container:
      image: quay.io/devfile/universal-developer-image:ubi8-latest
----
====
+
TIP: For more details, see the link:https://devfile.io/docs/2.2.0/what-is-a-devfile[devfile v2 documentation].

. Create a `DevWorkspace` custom resource, pasting the devfile contents from the previous step under the `spec.template` field.
+
.A `DevWorkspace` custom resource
====
[source,yaml,subs="+quotes,+attributes"]
----
kind: DevWorkspace
apiVersion: workspace.devfile.io/v1alpha2
metadata:
  name: my-devworkspace#<1>
  namespace: user1-dev#<2>
spec:
  routingClass: che
  started: true#<3>
  contributions:#<4>
    - name: ide
      kubernetes:
        name: che-code#<5>
  template:
    projects:#<6>
      - name: my-project-name
        git:
          remotes:
            origin: https://github.com/eclipse-che/che-docs
    components:#<7>
      - name: tooling-container
        container:
          image: quay.io/devfile/universal-developer-image:ubi8-latest
----

> <1> Name of the `DevWorkspace` custom resource. This is will be the name of the new workspace.

Reviewer 2: ```suggestion
<1> Name of the `DevWorkspace` custom resource. This will be the name of the new workspace.
```

[id="creating-workspaces"]
= Creating workspaces

If your use case does not permit use of the {prod-short} dashboard, you can create workspaces with {orch-name} APIs by applying custom resources to the cluster.

[NOTE]
====

Creating new workspaces through the {prod-short} dashboard provides better user experience and configuration benefits compared to using the command line:

* As a user, you are automatically logged in to the {orch-name} cluster.
* {platforms-name} clients work automatically.
* {prod-short} and its components automatically convert the target Git repository's devfile into the `DevWorkspace` and `DevWorkspaceTemplate` custom resources on the cluster.
* Access to the workspace is secured by default with the `routingClass: che` in the `DevWorkspace` of the workspace.

> * Recognition of configurations in `spec.devEnvironments` specified in the `CheCluster` custom resource including:
> ** Persistent storage strategy specified with `devEnvironments.storage`
> ** Default IDE specified with `devEnvironments.defaultEditor`
> ** Default plugins specified with `devEnvironments.defaultPlugins`
> ** Container build configuration specified with `devEnvironments.containerBuildConfiguration`
> * Recognition of the `DevWorkspaceOperatorConfig` configuration managed by {prod-short}

Reviewer: ```suggestion
* Recognition of the `DevWorkspaceOperatorConfig` configuration is managed by {prod-short}.
* Recognition of configurations in `spec.devEnvironments`is specified in the `CheCluster` custom resource including:
** Persistent storage strategy is specified with `devEnvironments.storage`
** Default IDE is specified with `devEnvironments.defaultEditor`
** Default plugins are specified with `devEnvironments.defaultPlugins`
** Container build configuration is specified with `devEnvironments.containerBuildConfiguration`
```
